compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/cfa/CfaUtils.kt

```diff
@@ -52,17 +53,91 @@ class LocalPropertyCollector private constructor() : ControlFlowGraphVisitorVoid
     }
 }
 
+class PathAwarePropertyInitializationInfo(
+    map: PersistentMap<String?, PropertyInitializationInfo> = persistentMapOf()
+) : ControlFlowInfo<PathAwarePropertyInitializationInfo, String?, PropertyInitializationInfo>(map) {
+    companion object {
+        val INIT = PathAwarePropertyInitializationInfo(persistentMapOf(null to PropertyInitializationInfo.EMPTY))
+    }
+
+    override val constructor: (PersistentMap<String?, PropertyInitializationInfo>) -> PathAwarePropertyInitializationInfo =
+        ::PathAwarePropertyInitializationInfo
+
+    fun applyLabel(node: CFGNode<*>, label: String?): PathAwarePropertyInitializationInfo {
+        if (label == null) {
+            // Special case: when we exit the try expression, null label means a normal path.
+            // Filter out any info bound to non-null label
+            // One day, if we allow multiple edges between nodes with different labels, e.g., labeling all paths in try/catch/finally,
+            // instead of this kind of special handling, proxy enter/exit nodes per label are preferred.
+            if (node is TryExpressionExitNode) {
+                return if (map.containsKey(null)) {
+                    constructor(persistentMapOf(null to map[null]!!))
+                } else {
+                    /* This means no info for normal path. */
+                    INIT
+                }
+            }
+            // In general, null label means no additional path info, hence return `this` as-is.
+            return this
+        }
+
+        val hasNonNullLabels = map.keys.any { it != null }
+        return if (hasNonNullLabels) {
+            // { |-> ... l1 |-> I1, l2 |-> I2, ... }
+            //   | l1         // path exit: if the given info has non-null labels, this acts like a filtering
+            // { |-> I1 }     // NB: remove the path info
+            if (map.keys.contains(label)) {
+                constructor(persistentMapOf(null to map[label]!!))
+            } else {
+                /* This means no info for the specific label. */
+                INIT
+            }
+        } else {
+            // { |-> ... }    // empty path info
+            //   | l1         // path entry
+            // { l1 -> ... }  // now, every info bound to the label
+            constructor(persistentMapOf(label to (map[null] ?: PropertyInitializationInfo.EMPTY)))
+        }
+    }
+
+    fun merge(other: PathAwarePropertyInitializationInfo): PathAwarePropertyInitializationInfo {
+        var resultMap = persistentMapOf<String?, PropertyInitializationInfo>()
+        for (label in keys.union(other.keys)) {
+            // disjoint merging to preserve paths. i.e., merge the property initialization info iff both have the key.
```

Comment:
Typo: `iff`

Reply:
It was intentional: https://en.wikipedia.org/wiki/If_and_only_if

if both `info`s have the same key, their corresponding property init info will be merged. If a certain key has a merged form of property init info, that means that key appears at both paths.

Anyway, I changed to explicitly say that term.